### PR TITLE
Detector train logging

### PR DESCRIPTION
**Pull request by VRehnberg**
This PR creates a TrainConfig base dataclass that is used jointly by both `train_classifier` script and `detector.train` methods that uses optimizers (finetuning and abstractions currently).

**N.B.** This is a breaking change as some of the command line arguments are changing so old train_classifier configs cannot be loaded without effort. I realised this a bit late, but shouldn't be too hard to avoid this if that is preferred.

Main value that is added is having better defaults and logging for train_detector.py but I might add validation for detector training as well here. Though, validation should probably be consistent regardless of how detector is trained (gradient-based or statistical).

### CODE REVIEW

**Comment by VRehnberg** [comment]
Might be some additional issues with this. It's tested with the test-suite and for some actual usage, but noticed some stuff that seemed like it could affect this when I was trying to add validation to this.

**Comment by ejnnr** [comment]
I like this in terms of design! Haven't taken a line-by-line look yet since it's marked as draft. Let me know when you'd like me to take a look (I might be slow to respond before Jan 2nd though).

**Comment by ejnnr** [comment]
Oh, and breaking changes are totally fine by me at this stage

**Comment by VRehnberg** [comment]
There are some other things that I was hoping to do in this commit as well:
 - Have a shared LightningModule base class for all detectors and refactor the detector training to be more straight forward
 - Introduce validation step for the detector training

But, it's a bit too much that would have to change for me to get a good grasp of it so I'm stashing it for now. After I found the finetuning bug (https://github.com/ejnnr/cupbearer/pull/26/commits/cbc87181ee40efc2331c289c72ba999b96fbda9d), I have less need of if anyway.

File: src/cupbearer/detectors/mahalanobis/__init__.py [hunk 1]
```diff
@@ -1,24 +1,26 @@
...
+from cupbearer.utils.utils import BaseConfig
 
 from .mahalanobis_detector import MahalanobisDetector
 
 
 @dataclass
-class MahalanobisTrainConfig(TrainConfig):
-    max_batches: int = 0
+class MahalanobisTrainConfig(BaseConfig):
+    max_batches = 0
```
Comment (ejnnr) [inline src/cupbearer/detectors/mahalanobis/__init__.py:11]: Was removing the type hint here deliberate?

**Review by ejnnr — COMMENTED** [review]
Left just a few minor notes, otherwise looks good, so feel free to merge this without another review after going over my comments. (But you can request one of course if changes end up being bigger and it seems helpful)

File: src/cupbearer/detectors/finetuning.py [hunk 2]
```diff
@@ -1,19 +1,14 @@
...
 from dataclasses import dataclass, field
-from typing import Optional
 
-import lightning as L
 import torch
 import torch.nn.functional as F
-from torch.utils.data import DataLoader
 
 from cupbearer.detectors.anomaly_detector import AnomalyDetector
 from cupbearer.detectors.config import DetectorConfig, TrainConfig
```
[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/detectors/finetuning.py:9]: It's still imported in `cupbearer.detectors.config` so still available, but better to import it from it's actual source directly. Good catch.

File: src/cupbearer/detectors/finetuning.py [hunk 3]
```diff
@@ -26,34 +21,22 @@ def __init__(self, model, max_batch_size, save_path):
...
-        optimizer: OptimizerConfig,
-        num_classes: int,
-        num_epochs: int = 10,
-        batch_size: int = 128,
-        max_steps: Optional[int] = None,
-        log_every_n_steps: Optional[int] = None,
-        **kwargs,
+        *,
+        num_classes,
+        train_config,
```
[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/detectors/finetuning.py:26]: Did both

[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/detectors/finetuning.py:26]: But, statistical detectors are using a different TrainConfig so the annotation in the ABC is just a BaseConfig (should perhaps have a different common ancestor (like DataLoaderConfig) instead that covers similarities).

File: src/cupbearer/detectors/mahalanobis/__init__.py (hunk 1 above)
Comment (VRehnberg) [reply src/cupbearer/detectors/mahalanobis/__init__.py]: Oops, don't know how that could have happened. I probably should start reviewing my own code as well...
